### DISCUSSION TITLE
Preloading / Prefetching IAsyncEnumerable

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#64697] [Q&A]
I have an IAsyncEnumerable that returns batches of items from a web service.
It takes ~5 seconds to return a batch and ~5 seconds to process a batch.
This means that it takes ~10 seconds for each batch to process.
However, this could be amortized down to ~5 seconds if it were possible for the batch enumerator to keep running in the background and store the items in buffer.  Is there an API that does this?

I'm hoping to do something like this:
```
var Query = ApiService.GetBatchAsyncEnumerable()
   .BufferUpTo(5) //This method
   .Select(x =>
```

**Comment by gfoidl** [comment]
You could use [Channels](https://devblogs.microsoft.com/dotnet/an-introduction-to-system-threading-channels/) for that.

**Reply by Xyncgas** [reply]
```C#
await foreach (var ele in MyChannel.Reader.ReadAllAsync())
{
}
```
This should do the job to read all the objects while the writer can put objects on its own (so it can put 5 items at once while u are currently processing 1 and then when you finished processing all the others are ready for processing)
